lib/lynis/lynistest.pm

```diff
@@ -0,0 +1,289 @@
+# Copyright (C) 2021 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+# Summary: Base module for Lynis test cases
+# Maintainer: llzhao <llzhao@suse.com>
+# Tags: poo#78224
+
+package lynis::lynistest;
+
+use strict;
+use warnings;
+use testapi;
+use utils;
+use autotest;
+
+use base 'consoletest';
+use LTP::TestInfo 'testinfo';
+
+our @EXPORT = qw(
+  $testdir
+  $f_position_b
+  $f_position_c
+  @lynis_ok
+  @lynis_fail
+  @lynis_softfail
+  $lynis_baseline_file
+  $lynis_baseline_file_default
+  $lynis_audit_system_current_file
+  $lynis_audit_system_error_file
+  parse_lynis_section_list
+  load_lynis_section_tests
+  parse_lynis_section_content
+  compare_lynis_section_content
+);
+
+our $testdir = "/tmp/";
+
+# File pointer
+our $f_position_b = 0;
+our $f_position_c = 0;
+
+our $lynis_baseline_file_default = "baseline-lynis-audit-system-nocolors-sle15sp3-x86_64-snapshot7-textmode";
+our $lynis_baseline_file         = get_var("LYNIS_BASELINE_FILE", $lynis_baseline_file_default);
+
+our $lynis_audit_system_current_file = "lynis_audit_system_current_file";
+our $lynis_audit_system_error_file   = "lynis_audit_system_error_file";
+
+# Lynix test status mappping with openQA
+our @lynis_ok       = split(/,/, get_var("LYNIS_OK",      "OK,DONE,YES"));
+our @lynis_fail     = split(/,/, get_var("LYNIS_ERROR",   "ERROR,WEAK,UNSAFE"));
+our @lynis_softfail = split(/,/, get_var("LYNIS_WARNING", "WARNING,EXPOSED,NONE,SUGGESTION"));
+
+sub loadtest_lynis {
+    my ($test, %args) = @_;
+    autotest::loadtest("lib/lynis/$test.pm", %args);
+}
+
+# Parse lynis test report sections
+# input: $file - file name
+# output: @section_list - section list
+sub parse_lynis_section_list {
+    my ($file) = @_;
+    # Section token
+    my $s_tok        = '\[\+\] ';
+    my @section_list = ();
+
+    my $rf;
+    my $i = 0;
+
+    # Parse the "sections" of input file
+    open($rf, $file) or die "Can not open $file, $!";
+    while (my $line = <$rf>) {
+        if ($line =~ /$s_tok/) {
+            push @section_list, $line;
+            $i++;
+        }
+    }
+    record_info("T: $i", "Total \"$i\" sections found in file \"$file\":\n @section_list");
+    close($rf) or die "Can not close file, $!";
```

Comment:
if it is "Can not close **$file**, $!";  here ?

Reply:
yes

> Going through the Verification Run, found the Networking section is Not Same, may I know the judgement rule for lynis,

Just compare the current outputs with baseline file.
Yes, section contents may changed. In this code I use `237         if ("join('', @section_baseline)" eq "join('', @section_current)") {` to compare the contents in lib file `lynistest`.

> when the comparison is different (not the same as baseline) and should be a bug or something else?
If not different the report a "warning" or "softfail" first then check the current output if there are any "$errors", if there are $errors then test case will report "failed".